hack/generated/pkg/util/armresourceresolver/resolver.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package armresourceresolver
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/Azure/k8s-infra/hack/generated/pkg/genruntime"
+	"github.com/Azure/k8s-infra/hack/generated/pkg/util/kubeclient"
+)
+
+type Resolver struct {
+	client *kubeclient.Client
+}
+
+func NewResolver(client *kubeclient.Client) *Resolver {
+	return &Resolver{
+		client: client,
+	}
+}
+
+// GetResourceGroupAndFullAzureName gets the full name for use in creating a resource. This name includes
+// the full "path" to the resource being deployed. For example, a Virtual Network Subnet's name might be:
+// "myvnet/mysubnet"
+func (r *Resolver) GetResourceGroupAndFullAzureName(ctx context.Context, obj genruntime.MetaObject) (string, string, error) {
+	owner := obj.Owner()
+
+	if obj.GetObjectKind().GroupVersionKind().Kind == "ResourceGroup" {
+		return getAzureName(obj), "", nil
+	}
+
+	if owner == nil {
+		return "", "", errors.New(fmt.Sprintf(
+			"Can't GetOwnerAndResourceGroupDetails from %s (kind: %s), which has no owner but is not a ResourceGroup",
+			obj.GetName(),
+			obj.GetObjectKind().GroupVersionKind()))
+	}
+
+	// TODO: This is a hack for now since we don't have an RG type yet
+	if owner.Kind == "ResourceGroup" {
+		return owner.Name, getAzureName(obj), nil
+	}
+
+	var ownerGvk schema.GroupVersionKind
+	found := false
+	// TODO: We need to find the specific storage version GVK...
+	for gvk := range r.client.Scheme.AllKnownTypes() {
+		if gvk.Group == owner.Group && gvk.Kind == owner.Kind {
+			if !found {
+				ownerGvk = gvk
+				found = true
+			} else {
+				return "", "", errors.Errorf("owner group: %s, kind: %s has multiple possible schemes registered", owner.Group, owner.Kind)
+			}
+		}
+	}
+
+	// TODO: We could do this on launch probably since we can check based on the AllKnownTypes() collection
+	if !found {
+		return "", "", errors.Errorf("couldn't find registered scheme for owner %+v", owner)
+	}
+
+	ownerNamespacedName := types.NamespacedName{
+		Namespace: obj.GetNamespace(), // TODO: Assumption that resource ownership is not cross namespace
```

Comment:
Do we need to raise some kind of error (or generate a log entry) if this is violated? Can we tell?

Reply:
`r.client.GetObject` below will fail if it can't find the resource. So the customer won't be able to link across namespace.

Additionally, they're unable to specify a namespace for their owner anyway.

Kubernetes also dissalows cross-namespace owner references: https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/

>Cross-namespace owner references are disallowed by design. This means:
>Namespace-scoped dependents can only specify owners in the same namespace, and owners that are cluster-scoped.
>Cluster-scoped dependents can only specify cluster-scoped owners, but not namespace-scoped owners.

I believe for us, since all of our resources are namespaced, there cannot be cluster-scoped owners (because you can't deploy a cluster scope ResourceGroup, BatchAccount, etc). So this is really probably not even a TODO, I've demoted it to just a normal comment (that references the GC doc).

Reply:
Fair enough.